skills_ml/algorithms/embedding/train.py

```diff
@@ -55,83 +57,110 @@ def __init__(
+        self.corpus_metadata = None
+
+    def _train_one_batch(self, model, batch, *args, **kwargs):
+        if len(model.wv.vocab) == 0:
+            model.build_vocab(batch, update=False)
+        else:
+            model.build_vocab(batch, update=True)
 
-    def train(self, lookup=False, *args, **kwargs):
+        model.train(batch, total_examples=model.corpus_count, epochs=model.iter, *args, **kwargs)
+        return model
+
+    def train(self, corpus_generator, n_processes=1, lookup=False, *args, **kwargs):
         """Train an embedding model, build a lookup table and model metadata. After training, they will be saved to S3.
 
         Args:
-            kwargs: all arguments that gensim.models will take.
+            corpus_generator (:generator): the iterable corpus
+            n_processes (:int): number of the processes
+            lookup (:bool): if True, the lookup dictionary of the corpus will be saved. It's more useful for doc2vec model
+            kwargs: all arguments that gensim.models.train will take.
         """
-        if self.model_type in ['word2vec', 'fasttext']:
-            if self._model.wv.vocab:
-                logging.info("Model has been trained")
-                self.update = True
-
-            batch_iter = 1
-            batch_gen = batches_generator(self.corpus_generator, self.batch_size)
-            for batch in batch_gen:
-                batch = Reiterable(batch)
-                logging.info("Training batch #{} ".format(batch_iter))
-                if not self.update:
-                    self._model.build_vocab(batch, update=False)
-                    self.update = True
-                else:
-                    self._model.build_vocab(batch, update=True)
-
-                self._model.train(batch, total_examples=self._model.corpus_count, epochs=self._model.iter, *args, **kwargs)
-                self.vocab_size_cumu.append(len(self._model.wv.vocab))
-                batch_iter += 1
-                logging.info('\n')
-
-        elif self.model_type == 'doc2vec':
-            corpus_gen = self.corpus_generator
-            reiter_corpus_gen = Reiterable(corpus_gen)
-            self._model.build_vocab(reiter_corpus_gen)
-            self._model.train(reiter_corpus_gen, total_examples=self._model.corpus_count, epochs=self._model.iter, *args, **kwargs)
+        tic = time()
+        try:
+            self.corpus_metadata = corpus_generator.metadata
+        except AttributeError:
+            self.corpus_metadata = None
+
+        if any([model.wv.vocab for model in self._models]):
+            logging.info("Model has been trained")
+            return 0
+
+        for model in self._models:
+            model.model_name = "_".join([
+                model.model_type,
+                self._model_hash(model, self.training_time, self.corpus_metadata)]
+            ) + '.model'
+
+        if self.model_type <= set(['word2vec', 'fasttext']):
```

Comment:
This method is kind of long, maybe splitting this big middle section into `_train_batches` and `_train_full_corpus` would work